src/main/java/org/apache/mesos/hdfs/Scheduler.java

```diff
@@ -572,21 +602,57 @@ public void run() {
       log.info(String.format("NameNodes: %s", persistentState.getNameNodes()));
       log.info(String.format("DataNodes: %s", persistentState.getDataNodes()));
 
-      LinkedHashMap<Protos.TaskID, Protos.TaskStatus> runningTasks = liveState.getRunningTasks();
       Collection<String> taskIds = persistentState.getAllTaskIds();
-      Collection<Protos.TaskID> runningTaskIds = runningTasks.keySet();
+      Set<String> runningTaskIds = liveState.getRunningTasks().keySet();
       for (String taskId : taskIds) {
-        boolean taskFound = false;
-        for (Protos.TaskID runningTaskId : runningTaskIds) {
-          if (runningTaskId.getValue().equals(taskId)) {
-            taskFound = true;
-          }
-        }
-        if (!taskFound) {
+        if (!runningTaskIds.contains(taskId))
           persistentState.removeTaskId(taskId);
-        }
       }
       reconciliationCompleted = true;
     }
   }
```

Comment:
The scheduler is getting really bloated. Can you separate these methods into a separate class. Also, can you add some kind of timeout, perhaps for 30 seconds, so this is not called every time there is a resource offer?  private RateLimiter reloadLimiter = RateLimiter.create(1 / 30.); // no more than once every 30s